AVVideoCaptureSource::setSessionSizeFrameRateAndZoom is exiting too early in case of different frame rate

youennf opened the pull request:
#### 777429a3054a57247989c4e333839404c02a7bb3
<pre>
AVVideoCaptureSource::setSessionSizeFrameRateAndZoom is exiting too early in case of different frame rate
<a href="https://rdar.apple.com/130112520">rdar://130112520</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=275952">https://bugs.webkit.org/show_bug.cgi?id=275952</a>

Reviewed by Eric Carlson.

We wew only checking that frame rate range is matching, not that the frame rate is matching.
The range is something like [1, 30] or [1, 60], so we need to do actual frame rate matching.
This issue was hidden as frame rate adaptation happens later in the pipeline, but this is not good from a CPU point of view.

We change areSettingsMatching() to check whether the frame rate we want is matching with the device active min and max frame rate.

Manually tested on iPhone 15 Pro.

* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::isFrameRateMatching):
(WebCore::AVVideoCaptureSource::areSettingsMatching const):
(WebCore::AVVideoCaptureSource::setSessionSizeFrameRateAndZoom):
(WebCore::isSameFrameRateRange): Deleted.

Canonical link: <a href="https://commits.webkit.org/280417@main">https://commits.webkit.org/280417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6959f9bb224dc33f798b337be868e580c063470

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56584 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35911 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9057 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60196 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7025 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58711 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43534 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7216 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45834 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4914 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58614 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33761 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48833 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26695 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30542 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6165 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6028 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6436 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61878 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/496 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6541 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53091 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/497 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48892 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53062 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12520 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/427 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31739 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32825 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33909 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32571 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -795,18 +795,24 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
+static bool isFrameRateMatching(double frameRate, AVCaptureDevice* device)
 {
-    return a.minFrameRate == b.minFrameRate && a.maxFrameRate == b.maxFrameRate && !PAL::CMTimeCompare(a.minFrameDuration, b.minFrameDuration) && !PAL::CMTimeCompare(a.maxFrameDuration, b.maxFrameDuration);
+    auto frameDuration = PAL::CMTimeMake(1, frameRate);
```
Nit: you can define this just before it is used

youennf replied:
done

youennf committed:
AVVideoCaptureSource::setSessionSizeFrameRateAndZoom is exiting too early in case of different frame rate
rdar://130112520
https://bugs.webkit.org/show_bug.cgi?id=275952

Reviewed by Eric Carlson.

We wew only checking that frame rate range is matching, not that the frame rate is matching.
The range is something like [1, 30] or [1, 60], so we need to do actual frame rate matching.
This issue was hidden as frame rate adaptation happens later in the pipeline, but this is not good from a CPU point of view.

We change areSettingsMatching() to check whether the frame rate we want is matching with the device active min and max frame rate.

Manually tested on iPhone 15 Pro.

* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::isFrameRateMatching):
(WebCore::AVVideoCaptureSource::areSettingsMatching const):
(WebCore::AVVideoCaptureSource::setSessionSizeFrameRateAndZoom):
(WebCore::isSameFrameRateRange): Deleted.

Canonical link: https://commits.webkit.org/280417@main
